### DISCUSSION TITLE
hasOne relationship with pivot table

### DISCUSSION THREAD

**Discussion by MiguelFreire** [adonisjs/core#3782] [Help]
Hello,

Let's suppose I have the following database schema:
```
users
id|email|password
```
```
avatars
id|name|url
```
```
user_avatars
id|user_id|avatar_id
```
How can I create an hasOne relationship with a pivot table?

I could solve this by adding a nullable foreign key column to users, but I don't really like that. So I was thinking of using an intermediate model (UserAvatar) to represent the pivot table (users_avatars) and set a belongsTo to the other two models, and in the User and Avatar model create a hasOne to UserAvatar model.

Any suggestions/ideas?
Ty

**Comment by Melchyore** [comment]
@MiguelFreire Hello!

Why do you want to create a pivot table for a 'hasOne' relationship? Pivot tables are meant to be used with "manyToMany" relationships.

Take a look at the [hasOne docs](https://docs.adonisjs.com/guides/models/relationships#hasone).

**Reply by MiguelFreire** [reply]
Basically, I don't want to add a nullable foreign key column to my users table (avatar_id) or vice-versa.
Also in my data layer Avatars and Users can exist by themselves, but users can use one of several existing avatars.

**Reply by galaczi** [reply]
It's probably a many to many relationship then.

Also, with hasOne relationship you would have the userId field on the avatars, not avatarId on users.

**Reply by Melchyore** [reply]
If a user can have ONLY one avatar and one avatar can belong to ONLY one user, then it's a has one relationship. If so, then you just have to add the user_id to the avatars table.

**Comment by thetutlage** [comment]
There are a couple of ways around it. 

- First, you can make a user belongs to an avatar (reads weird), but that way, the `avatars` table will have the `user_id`.
- Another one is to use [attachment-lite](https://github.com/adonisjs/attachment-lite). It stores attachments (in your use avatar) as a JSON blob within the same table
